crates/proof-of-sql-planner/src/plan.rs

```diff
@@ -0,0 +1,472 @@
+use super::{expr_to_proof_expr, table_reference_to_table_ref, PlannerError, PlannerResult};
+use alloc::vec::Vec;
+use datafusion::{
+    common::DFSchema,
+    logical_expr::{Expr, LogicalPlan, TableScan},
+    sql::{sqlparser::ast::Ident, TableReference},
+};
+use indexmap::IndexMap;
+use proof_of_sql::{
+    base::database::{ColumnField, ColumnRef, ColumnType, TableRef},
+    sql::{
+        proof_exprs::{AliasedDynProofExpr, DynProofExpr, ProofExpr, TableExpr},
+        proof_plans::DynProofPlan,
+    },
+};
+
+/// Get `AliasedDynProofExpr` from a `TableRef`, column indices for projection as well as
+/// input and output schemas
+///
+/// Note that at least in the current implementation of `DataFusion`
+/// the output schema should be a subset of the input schema
+/// and that no aliasing should take place.
+/// However that shouldn't be taken for granted.
+fn get_aliased_dyn_proof_exprs(
+    table_ref: &TableRef,
+    projection: Option<Vec<usize>>,
+    input_schema: &DFSchema,
+    output_schema: &DFSchema,
+) -> PlannerResult<Vec<AliasedDynProofExpr>> {
+    let num_input_columns = input_schema.columns().len();
+    let projection_indexes =
+        projection.unwrap_or_else(|| (0..num_input_columns).collect::<Vec<_>>());
+    projection_indexes
+        .iter()
+        .enumerate()
+        .map(
+            |(output_index, input_index)| -> PlannerResult<AliasedDynProofExpr> {
+                // Get output column name / alias
+                let alias: Ident = output_schema.field(output_index).name().as_str().into();
+                let input_column_name: Ident =
+                    input_schema.field(*input_index).name().as_str().into();
+                let data_type = input_schema.field(*input_index).data_type();
+                let expr = DynProofExpr::new_column(ColumnRef::new(
+                    table_ref.clone(),
+                    input_column_name,
+                    ColumnType::try_from(data_type.clone()).map_err(|_e| {
+                        PlannerError::UnsupportedDataType {
+                            data_type: data_type.clone(),
+                        }
+                    })?,
+                ));
+                Ok(AliasedDynProofExpr { expr, alias })
+            },
+        )
+        .collect::<PlannerResult<Vec<_>>>()
+}
+
+/// Convert a `TableScan` without filters or fetch limit to a `DynProofPlan`
+fn table_scan_to_projection(
+    table_name: &TableReference,
+    schemas: &IndexMap<TableReference, DFSchema>,
+    projection: Option<Vec<usize>>,
```

Comment:
Passing as a reference will avoid clones, and it is only ever iterated over, so ownership does nothing.
```suggestion
    projection: &Option<Vec<usize>>,
```